platforms/jvm/plugins-java-application/src/main/java/org/gradle/api/plugins/JavaApplicationPlugin.java

```diff
@@ -59,7 +59,8 @@
  *
  * @see <a href="https://docs.gradle.org/current/userguide/application_plugin.html">Application plugin reference</a>
  */
-public abstract class ApplicationPlugin implements Plugin<Project> {
+@SuppressWarnings("deprecation")
+public abstract class JavaApplicationPlugin extends ApplicationPlugin implements Plugin<Project> {
```

Comment:
❌ I think `PluginContainer` also doesn't expect plugins to extend one another. 

It looks to me that `findPlugin(Class)` and `hasPlugin(Class)` rely on the class being exactly the same vs in the hierarchy. That means `plugins.hasPlugin(ApplicationPlugin)` would be true before this change and false after this change with the same plugin applied.